core-libs/setup/ssr/engine/cx-common-engine-wrapper.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { StaticProvider } from '@angular/core';
+import { CommonEngine } from '@angular/ssr';
+import { CxServerErrorResponse } from '../error-handling/server-error-response';
+import { PROPAGATE_SERVER_ERROR_RESPONSE } from '../error-handling/server-error-response/propagate-server-error-response';
+import { RenderOptions } from './ng-express-engine';
+
+/**
+ * Wraps the CommonEngine render method to handle propagated server responses caught during server-side rendering.
+ */
+
+export function cxCommonEngineWrapper(
+  engine: CommonEngine,
+  renderOptions: RenderOptions,
+  callback: (err?: Error | null, html?: string) => void
+) {
```

Comment:
IMHO this method should not contain `callback`. IMHO our we should only build a wrapper over the `CommonEngine.render` method, but the `callback` interface is an implementation detail of the calling class which is the `ngExpressEngine`.

Please note, that theoretically in the future we might get rid of the `ngExpressEngine` (e.g. if we go serverless), but what will last is our wrapper over the `CommonEngine.render` method which is handling errors. 


What about creating a child class of CommonEngine, named e.g. `CxCommonEngine`. In the child class we could override the method `render()` inside which we could call the `super.render()`.

usage: in ngExpressEngine:
```ts
const commonEngine = new CxCommonEngine(commonEngineOptions);
/*...*/
commonEngine.render(renderOptions)
  .then(html => callback(null, html)
  .catch(callback)
```

Please note that such an approach allows also for using our child class also in non-callback-ish environment, e.g. in a pure `server.ts`

```ts
const commonEngine = new CxCommonEngine(commonEngineOptions);
/* ... */
       server.get('*', (req, res, next) => {
          commonEngine
            .render({
              bootstrap: AppServerModule,
              documentFilePath: indexHtml,
              url: req.originalUrl,
              publicPath: distFolder,
            })
            .then((html) => res.send(html))
            .catch((err) => next(err));
        });
```

Reply:
TBH I'm not sure the "serverless"  approach will be every used in conjunstion with this piece of code. I mentioned it just as an example of different circumstances in which the good separation of concerns becomes more clear to see. 

i.e. The Extended CommonEngine should only care about what extra functionality it provides (i.e. capturing async errors and populating them up as if it was a native feature of the CommonEngine). But it should not care about the context it is used withing. For example, the CommonEngine can be used inside a NgExpressEngine OR inside server.ts.